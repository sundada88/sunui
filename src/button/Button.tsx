import { defineComponent, PropType } from 'vue'
import './button.less'
import '../styles/common.less'
import { props } from './props'
import ripple from '../ripple/index'

export default defineComponent({
  name: 'SunButton',
  directives: {
    ripple,
  },
  props,
  // 定义的 props 属性 onClick 在编译的时候，会编译成 @click 事件，我们在外面监听事件可以通过 @click 监听
  /* 
    <button @click="handleClick"></button>
    <button :onClick="handleClick"></button>
    以上两种方法有相同的编译结果
  */
  setup(props, { slots }) {
    console.log('button setup')
    const handleClick = (e: Event) => {
      const { loading, disabled, onClick } = props
      if (loading || disabled) return
      onClick?.(e)
    }
    const handleTouchstart = (e: Event) => {
      const { loading, disabled, onTouchstart } = props
      if (loading || disabled) return
      onTouchstart?.(e)
    }

    return () => {
      const { text, color, type, block, size, outline, disabled, textColor, round } = props
      return (
        <button
          class={[
            'sun-button',
            'sun--box',
            `sun-button--${size}`,
            disabled ? 'sun-button--disabled' : null,
            text ? `sun-button--text-${type}` : `sun-button--${type}`,
            text && disabled ? 'sun-button--text-disabled' : null,
            block ? 'sun--flex sun-button--block' : 'sun--inline-flex',
            outline ? 'sun-button--outline' : null,
            text ? 'sun-button--text' : 'sun-elevation--1',
            round ? 'sun-button--round' : null,
          ]}
          style={{ background: color, color: textColor }}
          onClick={handleClick}
          onTouchstart={handleTouchstart}
          v-ripple={ripple}
        >
          {slots.default?.()}
        </button>
      )
    }
  },
})
