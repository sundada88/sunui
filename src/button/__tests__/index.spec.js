// import example from '../example'
import Button from '..'
import SunButton from '../Button'
import { createApp } from 'vue'
import { mount } from '@vue/test-utils'

test('test button plugin', () => {
  const app = createApp({}).use(Button)
  expect(app.component(Button.name)).toBeTruthy()
})

test('test button onClick && onTouchstart null callback', () => {
  const wrapper = mount(SunButton, {
    props: {
      onClick (e) {}
    }
  })
  wrapper.trigger('click')
  wrapper.trigger('touchstart')
  expect(wrapper.emitted()).toHaveProperty('click')
  expect(wrapper.emitted()).toHaveProperty('touchstart')
  wrapper.unmount()
})

test('test button onClick && onTouchstart', () => {
  const onClick = jest.fn()
  const onTouchstart = jest.fn()

  const wrapper = mount(SunButton, {
    props: {
      onClick,
      onTouchstart
    }
  })

  wrapper.trigger('click')
  wrapper.trigger('touchstart')
  expect(onClick).toHaveBeenCalledTimes(1)
  expect(onTouchstart).toHaveBeenCalledTimes(1)
  wrapper.unmount()
})

test('test button disabled', () => {
  const onClick = jest.fn()
  const onTouchstart = jest.fn()
  const wrapper = mount(SunButton, {
    props: {
      onClick,
      onTouchstart,
      disabled: true
    }
  })
  wrapper.trigger('click')
  wrapper.trigger('touchstart')
  expect(wrapper.classes()).toContain('sun-button--disabled')
  expect(onClick).toHaveBeenCalledTimes(0)
  expect(onTouchstart).toHaveBeenCalledTimes(0)
  // expect(wrapper.html()).toMatchSnapshots()
  wrapper.unmount()
})

test('test button color & textColor', () => {
  const wrapper = mount(SunButton, {
    props: {
      color: '#000',
      textColor: '#000'
    }
  })
  expect(wrapper.attributes('style')).toMatch('color: rgb(0, 0, 0)')
  expect(wrapper.attributes('style')).toMatch('background: rgb(0, 0, 0)')
  // expect(wrapper.html()).toMatchSnapshots()
  wrapper.unmount()
})

test('test button type', () => {
  const _default = 'button text'
  const wrapper = mount(SunButton, {
    props: {
      size: 'large',
      round: true,
      text: true,
      block: true
    },
    slots: {
      default: _default
    }
  })
  expect(wrapper.classes()).toContain('sun-button--large')
  expect(wrapper.classes()).toContain('sun-button--round')
  expect(wrapper.classes()).toContain('sun-button--text')
  expect(wrapper.classes()).toContain('sun-button--block')
  expect(wrapper.text()).toBe(_default)
  expect(wrapper.html()).toMatchSnapshot()
  wrapper.unmount()
})

test('test button text && disabled', () => {
  const wrapper = mount(SunButton, {
    props: {
      disabled: true,
      text: true
    }
  })
  expect(wrapper.classes('sun-button--text-disabled')).toBeTruthy()
  wrapper.unmount()
})

test('test button outline', () => {
  const wrapper = mount(SunButton, {
    props: {
      outline: true
    }
  })
  expect(wrapper.classes('sun-button--outline')).toBeTruthy()
  wrapper.unmount()
})
