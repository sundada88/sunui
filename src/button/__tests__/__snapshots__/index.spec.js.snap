// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`test button type 1`] = `"<button class=\\"sun-button sun--box sun-button--large sun-button--text-default sun--flex sun-button--block sun-button--text sun-button--round\\">button text</button>"`;
